# Atividade 1 — Exercícios com os comandos básicos de pandas para EDA


Responda as perguntas abaixo no notebook (cada resposta deve ter ao menos uma célula de código que produza o output e, logo abaixo, uma célula de texto com a sua interpretação).

## 1) Carregue o dataset e mostre as primeiras 10 linhas.

In [3]:
import pandas as pd
df = pd.read_csv('/content/emendas.csv')
df.head(10)

/tmp/ipykernel_8215/420280567.py:2: DtypeWarning: Columns (0,3,5,12) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('/content/emendas.csv')


,Código da Emenda,Ano da Emenda,Tipo de Emenda,Código do Autor da Emenda,Nome do Autor da Emenda,Número da emenda,Localidade de aplicação do recurso,Código Município IBGE,Município,Código UF IBGE,...,Código Ação,Nome Ação,Código Plano Orçamentário,Nome Plano Orçamentário,Valor Empenhado,Valor Liquidado,Valor Pago,Valor Restos A Pagar Inscritos,Valor Restos A Pagar Cancelados,Valor Restos A Pagar Pagos
0,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,ALVORADA - RS,4300604,ALVORADA,4300000.0,...,1D73,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,0000,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,"700000,00","0,00","0,00","0,00","690900,00","9100,00"
1,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,SANTA ISABEL DO PARÁ - PA,1506500,SANTA ISABEL DO PARÁ,1500000.0,...,10V0,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA,0000,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA -...,"487500,00","0,00","0,00","0,00","487500,00","0,00"
2,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,BARRA BONITA - SC,4202099,BARRA BONITA,4200000.0,...,8581,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0000,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,"100000,00","0,00","0,00","0,00","0,00","100000,00"
3,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,ARARAQUARA - SP,3503208,ARARAQUARA,3500000.0,...,10V0,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA,0000,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA -...,"243750,00","0,00","0,00","0,00","243750,00","0,00"
4,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,CAMPO LARGO - PR,4104204,CAMPO LARGO,4100000.0,...,8581,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0000,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,"1658333,94","0,00","0,00","0,00","527200,00","1131133,94"
5,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,CAMPINAS DO SUL - RS,4303806,CAMPINAS DO SUL,4300000.0,...,10V0,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA,0000,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA -...,"243750,00","0,00","0,00","0,00","0,00","243750,00"
6,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,MORENO - PE,2609402,MORENO,2600000.0,...,20RP,APOIO A INFRAESTRUTURA PARA A EDUCACAO BASICA,0000,INFRAESTRUTURA PARA A EDUCACAO BASICA - DESPES...,"700000,00","0,00","0,00","0,00","700000,00","0,00"
7,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,NOVA FLORESTA - PB,2510105,NOVA FLORESTA,2500000.0,...,8581,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0000,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,"278700,00","0,00","0,00","0,00","0,00","278700,00"
8,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,SANTIAGO DO SUL - SC,4215695,SANTIAGO DO SUL,4200000.0,...,20ZV,FOMENTO AO SETOR AGROPECUARIO,0000,FOMENTO AO SETOR AGROPECUARIO - DESPESAS DIVERSAS,"300000,00","300000,00","300000,00","0,00","0,00","0,00"
9,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,CIANORTE - PR,4105508,CIANORTE,4100000.0,...,1D73,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,0000,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,"245850,00","0,00","0,00","0,00","0,00","245850,00"


Sem comentários nessa primeira questão, vimos apenas o comando para importar e ver as primeiras 10 linhas do arquivo.

Explique em uma frase o que você observa nas primeiras 10 linhas.

## 2) Quantas linhas e colunas o dataset possui?

In [4]:
df.shape


(36371, 28)

O arquivo possui 36371 linhas e 28 colunas.

Explique o que esses números significam.

## 3) Liste os nomes das colunas e descreva brevemente o que você entende que cada coluna representa.

In [5]:
df.columns

Index(['Código da Emenda', 'Ano da Emenda', 'Tipo de Emenda',
       'Código do Autor da Emenda', 'Nome do Autor da Emenda',
       'Número da emenda', 'Localidade de aplicação do recurso',
       'Código Município IBGE', 'Município', 'Código UF IBGE', 'UF', 'Região',
       'Código Função', 'Nome Função', 'Código Subfunção', 'Nome Subfunção',
       'Código Programa', 'Nome Programa', 'Código Ação', 'Nome Ação',
       'Código Plano Orçamentário', 'Nome Plano Orçamentário',
       'Valor Empenhado', 'Valor Liquidado', 'Valor Pago',
       'Valor Restos A Pagar Inscritos', 'Valor Restos A Pagar Cancelados',
       'Valor Restos A Pagar Pagos'],
      dtype='object')

Ao todo, o arquivo possui 28 colunas que simbolizam aspectos da emenda (código, ano, tipo, autor, número da emenda) , aspectos do local de aplicação do recurso (município , UF, região) , aspectos da função (nome, código da subfunção, nome da subfunção) , aspectos do programa (código do programa, nome do programa) , aspectos de plano orçamentário (nome, valor empenhado, valor liquidado, valor pago, restos a pagar dos inscritos e dos cancelados, resto a pagar que foram pagos)

Escreva sua breve descrição das colunas, como se estivesse criando um dicionário de variáveis para que leitor entenda.

## 4) Existem valores nulos? quantos?


In [11]:
nulos = df.isnull().sum()
print(nulos)

Código da Emenda                      0
Ano da Emenda                         0
Tipo de Emenda                        0
Código do Autor da Emenda             0
Nome do Autor da Emenda               0
Número da emenda                      0
Localidade de aplicação do recurso    0
Código Município IBGE                 0
Município                             1
Código UF IBGE                        1
UF                                    1
Região                                1
Código Função                         1
Nome Função                           1
Código Subfunção                      1
Nome Subfunção                        1
Código Programa                       1
Nome Programa                         1
Código Ação                           1
Nome Ação                             1
Código Plano Orçamentário             1
Nome Plano Orçamentário               1
Valor Empenhado                       1
Valor Liquidado                       1
Valor Pago                            1


É visível a presença de valores nulos em alguns campos do arquivo, em colunas específicas mas sem oscilações extremas. Ou não tem valor nulo na coluna ou tem, no máximo, uma coluna.

## 5) Mostre os tipos das variáveis (se int, string, boolean, float...) e analise se alguma variável parece ter o tipo errado. Faça isso da maneira mais objetiva e otimizada possível.

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36371 entries, 0 to 36370
Data columns (total 28 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   Código da Emenda                    36371 non-null  object 
 1   Ano da Emenda                       36371 non-null  int64  
 2   Tipo de Emenda                      36371 non-null  object 
 3   Código do Autor da Emenda           36371 non-null  object 
 4   Nome do Autor da Emenda             36371 non-null  object 
 5   Número da emenda                    36371 non-null  object 
 6   Localidade de aplicação do recurso  36371 non-null  object 
 7   Código Município IBGE               36371 non-null  object 
 8   Município                           36370 non-null  object 
 9   Código UF IBGE                      36370 non-null  float64
 10  UF                                  36370 non-null  object 
 11  Região                              36370

A maioria das variáveis está corretamente tipada, com campos em object e 'Ano da Emenda' como int64, porém existem inconsistências importantes:'Código UF IBGE' aparece como float64 quando deveria ser int, muito provável que seja por existirem valores nulos, e todas as variáveis de valor financeiro estão em object mas deveria ser float, oque mostra problemas na formatação dos daods, como uso de vírgulas ou símbolos monetários.

Se alguma coluna estiver com o tipo errado, escreva abaixo qual é e por que isso pode atrapalhar análises.

**Dica:** Se alguma variável tiver no tipo o nome "object", isso se refere a texto.

## 6) Usando a coluna `Ano da Emenda` como base, descubra até que ano o dataset está atualizado (mostre as 8 primeiras linhas do df ordenado desc).

In [7]:
df_sorted = df.sort_values(by='Ano da Emenda', ascending=False)
df_sorted.head(8)

,Código da Emenda,Ano da Emenda,Tipo de Emenda,Código do Autor da Emenda,Nome do Autor da Emenda,Número da emenda,Localidade de aplicação do recurso,Código Município IBGE,Município,Código UF IBGE,...,Código Ação,Nome Ação,Código Plano Orçamentário,Nome Plano Orçamentário,Valor Empenhado,Valor Liquidado,Valor Pago,Valor Restos A Pagar Inscritos,Valor Restos A Pagar Cancelados,Valor Restos A Pagar Pagos
36340,201815040008,2018,Emenda Individual - Transferências com Finalid...,1504,JULIO LOPES,8,ITAPERUNA - RJ,3302205,ITAPERUNA,3300000.0,...,20UD,PREVENCAO A VIOLENCIA E A CRIMINALIDADE,EIND,EMENDA INDIVIDUAL,"266506,67","0,00","0,00","0,00","266506,67","0,00"
36354,201836980005,2018,Emenda Individual - Transferências com Finalid...,3698,JUNIOR MARRECA,5,MÚLTIPLO,Sem informação,Múltiplo,2100000.0,...,4525,APOIO A MANUTENCAO DE UNIDADES DE SAUDE,EIND,EMENDA INDIVIDUAL,"4039271,00","4039271,00","4039271,00","0,00","0,00","0,00"
36370,201836980022,2018,Emenda Individual - Transferências com Finalid...,3698,JUNIOR MARRECA,22,SENADOR ALEXANDRE COSTA - MA,2111748,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
36369,201836980021,2018,Emenda Individual - Transferências com Finalid...,3698,JUNIOR MARRECA,21,SÃO FÉLIX DE BALSAS - MA,2110807,SÃO FÉLIX DE BALSAS,2100000.0,...,10GD,"IMPLANTACAO, AMPLIACAO E MELHORIA DE SISTEMAS ...",EIND,EMENDA INDIVIDUAL,"600000,00","0,00","0,00","0,00","2823,86","597176,14"
36353,201836980004,2018,Emenda Individual - Transferências com Finalid...,3698,JUNIOR MARRECA,4,FEIRA NOVA DO MARANHÃO - MA,2104073,FEIRA NOVA DO MARANHÃO,2100000.0,...,4525,APOIO A MANUTENCAO DE UNIDADES DE SAUDE,EIND,EMENDA INDIVIDUAL,"140000,00","140000,00","140000,00","0,00","0,00","0,00"
36352,201836980003,2018,Emenda Individual - Transferências com Finalid...,3698,JUNIOR MARRECA,3,BERNARDO DO MEARIM - MA,2101939,BERNARDO DO MEARIM,2100000.0,...,4525,APOIO A MANUTENCAO DE UNIDADES DE SAUDE,EIND,EMENDA INDIVIDUAL,"500000,00","500000,00","500000,00","0,00","0,00","0,00"
36351,201836980002,2018,Emenda Individual - Transferências com Finalid...,3698,JUNIOR MARRECA,2,ARAME - MA,2100956,ARAME,2100000.0,...,4525,APOIO A MANUTENCAO DE UNIDADES DE SAUDE,EIND,EMENDA INDIVIDUAL,"500000,00","500000,00","500000,00","0,00","0,00","0,00"
36350,201836980001,2018,Emenda Individual - Transferências com Finalid...,3698,JUNIOR MARRECA,1,MÚLTIPLO,Sem informação,Múltiplo,2100000.0,...,14U2,"IMPLANTACAO, INSTALACAO E AMPLIACAO DE ESPACOS...",EIND,EMENDA INDIVIDUAL,"199999,98","1400,00","1400,00","11304,05","0,00","196456,36"


Como o comando dado foi de intuito aleatório, o último ano que aparece no meu google colab é o ano de 2018, embora na máquina de outros usuários o último ano de atualização seja 2025.

Explique: qual é o ano mais recente presente no dataset?

## 7) execute um  `describe()` e explique uma conclusão simples que você consegue tirar a partir das estatísticas (se algumas colunas deveriam aparecer no describe mas não aparecem, indique o motivo que você acredita que esteja causando esse problema).

In [8]:
df.describe()

,Ano da Emenda,Código UF IBGE
count,36371.000000,3.637000e+04
mean,2015.759424,3.079555e+06
std,1.398544,1.129218e+06
min,2014.000000,-1.400000e+01
25%,2014.000000,2.400000e+06
50%,2016.000000,3.100000e+06
75%,2017.000000,4.100000e+06
max,2018.000000,5.300000e+06


Acredito que o motivo pelo qual outras colunas não aparecem no describe() é que, por padrão, apenas aparecem colunas numéricas, deixando de lado as colunas do tipo objeto. Em suma, é visível as estastícas apenas das colunas numéricas 'Ano da Emenda' e 'Código UF IBGE'. Para 'Ano da Emenda', os ados vão de 2014 a 2018, indicando um dataset concentrado nesses anos.

## 8) Mostre as últimas 5 linhas do dataset e escreva se a última linha parece completa (sem muitos nulos) ou não.

In [9]:
df.tail()

,Código da Emenda,Ano da Emenda,Tipo de Emenda,Código do Autor da Emenda,Nome do Autor da Emenda,Número da emenda,Localidade de aplicação do recurso,Código Município IBGE,Município,Código UF IBGE,...,Código Ação,Nome Ação,Código Plano Orçamentário,Nome Plano Orçamentário,Valor Empenhado,Valor Liquidado,Valor Pago,Valor Restos A Pagar Inscritos,Valor Restos A Pagar Cancelados,Valor Restos A Pagar Pagos
36366,201836980018,2018,Emenda Individual - Transferências com Finalid...,3698,JUNIOR MARRECA,18,ITAPECURU MIRIM - MA,2105401,ITAPECURU MIRIM,2100000.0,...,20ZV,FOMENTO AO SETOR AGROPECUARIO,EIND,EMENDA INDIVIDUAL,"588056,05","5492,56","5492,56","17207,44","0,00","582563,49"
36367,201836980019,2018,Emenda Individual - Transferências com Finalid...,3698,JUNIOR MARRECA,19,ITAPECURU MIRIM - MA,2105401,ITAPECURU MIRIM,2100000.0,...,14UF,"CONSTRUCAO, REFORMA, EQUIPAGEM E AMPLIACAO DE ...",EIND,EMENDA INDIVIDUAL,"109663,98","0,00","0,00","2950,98","0,00","106713,00"
36368,201836980020,2018,Emenda Individual - Transferências com Finalid...,3698,JUNIOR MARRECA,20,MARANHÃO (UF),Sem informação,Sem informação,2100000.0,...,215F,FOMENTO E FORTALECIMENTO DA ECONOMIA SOLIDARIA...,EIND,EMENDA INDIVIDUAL,"500000,00","0,00","0,00","0,00","500000,00","0,00"
36369,201836980021,2018,Emenda Individual - Transferências com Finalid...,3698,JUNIOR MARRECA,21,SÃO FÉLIX DE BALSAS - MA,2110807,SÃO FÉLIX DE BALSAS,2100000.0,...,10GD,"IMPLANTACAO, AMPLIACAO E MELHORIA DE SISTEMAS ...",EIND,EMENDA INDIVIDUAL,"600000,00","0,00","0,00","0,00","2823,86","597176,14"
36370,201836980022,2018,Emenda Individual - Transferências com Finalid...,3698,JUNIOR MARRECA,22,SENADOR ALEXANDRE COSTA - MA,2111748,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


A última linha definitivamente não parece completa. Vários campos da última linha parecem estar incompletas, onde aparece apenas as letras "NaN".

Interprete a completude da última linha.

## 9) existem valores duplicados no dataframe? Caso positivo, quantos?

In [13]:
duplicated = df.duplicated().sum()
print(duplicated)

0


No dataframe aprasentado, não aparece nenhum valor duplicado.